Masonry backend (#1)
linebender/rfcs: rfcs/0001-masonry-backend.md

- There are [TODOs without an associated issue](https://github.com/linebender/xilem/blob/ea45b9f8c14e3708f0fcbe0a0e1c760f59146323/src/widget/core.rs#L569C5-L569C66).

While Masonry's codebase hasn't been updated in a while, I made sure to address the above problems before I stepped away from it. I would also argue it has a tighter vision and more helpful documentation, thought that is more subjective.

Masonry also comes with some built-in perks, like powerful unit tests and a structured widget graph.

For these reasons, I think a new backend should be added to Xilem using either the Masonry crate or one forked from it. (We will refer to that new backend as "Masonry" in the rest of this RFC, although it might have a different name in the end.)

## User-facing explanation

### Xilem architecture

Xilem is roughly separated into two layers:

- **The reactive layer** handles the app logic and centralized state. Your application's code will call functions from this layer.
- **The platform layer** handles the document state, depending on the platform.

> The reactive layer defines the `View` trait as well as other items for sequences, type-erasure and memoization. A `View` represent a transient node or tree of graphical items, which is return by a declarative function we usually call a "component". An example of component could be:

Reviewer: ```suggestion
The reactive layer defines the `View` trait as well as other items such as sequences of views, type-erasure and memoization. A `View` represents a transient description of one or multiple graphical items, which is usually returned by a declarative function that is often called a "component". An example of component could be:
```

Reviewer: Not really resolved, surely this is a little bit of nitpicking... but I don't see why the suggestion is not taken (as it should be a little bit more precise). And yes, I don't think the whole explanation of Xilems reactive architecture is necessary in this RFC (but since you've already written it, we can at least make it more accurate).

Author: > I don't see why the suggestion is not taken

The current text in the RFC:

> A `View` represents a transient description of one or multiple graphical items, which is usually returned by a declarative function that is often called a "component".

Is taken directly from your wording. I didn't keep the "such as sequences of views, type-erasure and memoization" part because the sentence "reactive layer defines [...] items such as [...] type-erasure and memoization" doesn't really parse right.

Reviewer: Well anyway, it's nitpicking, I'm fine with the current wording

```rust
fn app_logic(data: &mut u32) -> impl View<u32, (), Element = impl Widget> {
    Column::new((
        Button::new(format!("count: {}", data), |data| *data += 1),
        Button::new("reset", |data| *data = 0),
    ))
}
```

In that example, `app_logic()` returns a `Column` view which contains two `Button` views. These store the minimum information to represent the concept of two buttons stored in a column.

> The `View` trait has an `Element` associated type, which depends on the platform layer, and represents a persisten UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.

Reviewer: ```suggestion
The `View` trait has an associated type `Element`, which depends on the platform layer, and represents a persistent UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.
```

Not sure whether it "depends on the platform layer" but is just the actual type of the UI Widget (The platform is rather defined by the runtime).
As you say yourself below, the `View` trait is a different type on each platform, requiring different `Widget` traits for the `Element` type
 
Might be nitpicking though...

Author: Yeah, I rushed that explanation a bit. Didn't want to have to document the entire xilem_core paradigm just to justify switching a backend.

(In web frameworks that use a virtual DOM, this process is sometimes known as "reconciliation".)

Xilem has multiple platform layers. `xilem_web` is based on the web DOM, whereas `xilem_native` is based on the Masonry crate, which provides a native widget tree and some Developer Experience features.

Note that, while the platform layer is sometimes referred to as a "backend", it's somewhat tightly coupled with the reactive layer.

`xilem_web`, `xilem_native` and any other platforms will each have their own version of the `View` trait, the `ViewSequence` trait, the `AnyView` trait, the `BoxedView` type, and so on.

### `xilem_native`

> The element type in `xilem_native` must implement `masonry::Widget`.

Reviewer 2: To confirm, `xilem_native` would not contain any `Widget` implementations?

Widget in Masonry is a trait that represent units of graphical composition, things like text and buttons and boxes and containers. To implement Widget, a type must:

- Be `'static`.
- Provide info about its children.
- Handle user events.
- Compute its layout.
- Draw itself in a paint target.

See the [Widget trait definition](https://github.com/PoignardAzur/masonry-rs/blob/3cfc2cddec05d48a31a9812bdda27792d336f03f/src/widget/widget.rs#L65-L199) for more details.

### Directory structure

The `https://github.com/linebender/xilem` repository has the following crates in its `crates/` directory:

- `xilem_core/`
- `xilem_web/`
- `xilem_native/`
- `masonry/`

> - `xilem/`, which would only be a placeholder crate pointing to `xilem_web/` or `xilem_native/`.

Reviewer 2: Pointing to here meaning:
a) A link in the readme, and no content
b) Re-exporting only `xilem_native`
c) Re-exporting `xilem_native` or `xilem_web`, depending on platform
?

Reviewer 2: There are tradeoffs here. My preference in the near term would be for `xilem` itself to be the crate which users depend on for the native backend, with no need to know about `xilem_native`.

However, that does constrain the longer-term web/native backend unification/abstraction, which I hope might happen, to not use `xilem` as the crate name, which would be unfortunate.

## Implementation strategy

The implementation will require several steps:

- Refactoring Masonry to use the same dependencies Xilem currently uses, especially Glazier, Vello, and Parley insted of druid-shell and Piet.
- Integrating AccessKit support.
- Moving the Masonry codebase to the Xilem repository, ideally in a way that preserves commit history.
- Porting the list of Github Issues in the masonry-rs repository to the Xilem repository.
- Creating a new `xilem_native` crate on parity with the current Xilem crate.
- Removing the current xilem crate.
- Publishing both `xilem_native` and the placeholder `xilem` crate on crates.io.

> - Sharing ownership of the crates with Raph Levien.

Reviewer 2: I'd like to avoid over-specialising on the release process, see e.g. [#linebender > Our Ideal Crate Release Process](https://xi.zulipchat.com/#narrow/stream/419691-linebender/topic/Our.20Ideal.20crate.20release.20processs).

Reviewer: Agree, the RFC feels like it should be multiple RFCs, as it's not just about masonry as a backend.

Author: Right. I'll move that part to "future possibilities".

## Drawbacks

### Layer separation

This move decouples the reactive layer of xilem from its platform layer by putting them in separate crates.

> This means some features in `xilem_native` may be harder to implement because they won't have access to the backend's private code.

Reviewer: I think this needs more detailed explanation, do you mean the whole runtime? Only the widgets? The xilem `Widget` trait?

I'm also not sure whether this should be delayed and just prototyped when deemed necessary (and based on that experience a new RFC or depending on the complexity just a PR opened), I don't see much benefit in splitting xilem in too many crates currently (mostly because of the orphan rule). AFAIU the widgets in masonry will be its own crate already anyway, so the likely biggest amount of code is already in a separate crate.

Reviewer: Not sure if this is really resolved is it? (same with the other comments) I don't see where this is explained in more detail in the new version. I can only guess, that the temporary `xilem_masonry` will contain the runtime (i.e. is the temporary merger) and this will be the new `xilem` at some point

Author: It depends on what you mean by "the whole runtime".

Broadly speaking, the RFC makes a distinction between two "layers", the reactive layer and the platform layer. These layers are fuzzy, but real: the main difference between xilem and xilem_web is that xilem_web's platform layer is built on top of the web DOM. The xilem crate also has `src/view/` and `src/widget/` that are symptoms of this separation too.

In xilem, the "platform layer" is basically "everything except the `view/` folder". This RFC proposes replacing all of that by masonry, which means what would be left would be the `view/` folder, plus glue code to interface with Masonry.

> I'm also not sure whether this should be delayed and just prototyped when deemed necessary (and based on that experience a new RFC or depending on the complexity just a PR opened), I don't see much benefit in splitting xilem in too many crates currently (mostly because of the orphan rule). AFAIU the widgets in masonry will be its own crate already anyway, so the likely biggest amount of code is already in a separate crate.

To be clear, in the native case, there would be only two crates: `xilem` and `masonry`. (Plus `xilem_core`, which defines macro that both `xilem` and `xilem_web` use).

I guess I should write a more complete explainer about this. It feels a bit annoying, because a lot of the complexity I'm being asked to explain and justify already exists in Xilem, but in any case documenting things is probably a good thing.

Author: I guess I could draw up a list of public items in `xilem` and split them between "frontend" and "backend". That would probably make the explanation a lot easier to grok.

Reviewer: Hmm reading again, it's *kinda* clear in the RFC (by defining reactive layer and platform layer), so sorry for being so persistent, but it's a bigger change...

You should probably include a lot what you've written here in the RFC itself, I think it clears up things a little bit.

Basically it boils down to this:
* Keep the reactive view layer (basically everything in `src/view/*`)  in xilem
* Move runtime and everything else into masonry (in this RFC platform layer)

I think we may want to have a third crate then that actually contains the runtime (`xilem` and another name for the reactive layer?). At least in the current form of xilem the separation between view <-> runtime is not that simple (as the runtime requires the `View` trait).

So that the dependency chain would look like this: runtime -> reactive/view -> widgets (masonry)

(Or keep the runtime + reactive layer in one crate, I think this would be the simplest way forward for now)

Author: Yeah, figuring out the composition root might be a bit tricky, but I don't think it should be included in the RFC.

> You should probably include a lot what you've written here in the RFC itself

Will do. I do need to make these things more explicit.

> So that the dependency chain would look like this: runtime -> reactive/view -> widgets (masonry)

Most of the "runtime" as it currently exists will be defined in Masonry. The rough idea is that, when you create a Masonry window, you give it a bunch of callbacks (or a trait impl or whatever) to be called on every paint, input event, etc. So the composition root of Xilem will be in the callbacks passed to Masonry's own composition root.

Reviewer: > Yeah, figuring out the composition root might be a bit tricky, but I don't think it should be included in the RFC.

I think this was the reason why things were unclear to me a little bit and thus the question (runtime vs reactive layer etc.), since it's interconnected above `widget/` and `view/` (mainly `App` and `AppTask`) and thus probably not cleanly separable between "platform" and "reactive" layer (or we have a different interpretation of "reactive" layer) with just 2 crates (`xilem` and `masonry`).

Sure a lot of the code in these modules can be laid out in masonry etc. but as you say details can be done in PRs (I think it will clear up when `xilem_masonry` is being implemented).

Author: I've added a new sub-section that mentions the source files and how they'll be affected, though what the composition root will look like is left a bit vague.

Let me know if you think it should be more clear/explicit!

Reviewer: Yeah I think continuing discussion makes more sense in the relevant parts of the RFC

### Separating at the crate level vs the repository level

The initial suggestion was to use Masonry as it currently exists, as a separate repository.

However, people have brought up that it would make the Xilem development process a lot more complex. Features that reached across crates (`xilem_core`, `xilem_native`, `masonry`) would need to be rolled out at coordinated pull requests between repositories.

### Separating at the crate level vs the module level

Masonry and Xilem's code could simply be merged in a single crate, under different modules.

The consensus was mostly that having Masonry as a separate crate would help keep Xilem "honest" and enforce separation of concerns.

More importantly, Masonry was built as a potential foundation for *multiple* native GUI crates, and it would be valuable to conserve that; therefore, while Masonry would still depend on crates from the Linebender ecosystem, it shouldn't be too tied to Xilem.

### Porting Xilem to Masonry vs Masonry to Xilem

In any merge between two projects, one has to be the base the other is grafted on.

This RFC was written under the assumption Xilem would move towards Masonry, but the opposite could be possible.

> That said, the Motivation section explains why Masonry is a healthier codebase to start from.

Reviewer: I expect to just take the best code in one or the other and just use that (and otherwise merge the good bits of the other).
Xilem is the frontfacing layer, so it depends on masonry otherwise?

Author: Xilem as it currently is has both frontend and backend code. The question how much of that backend code we want to build from vs throwing it away and starting from Masonry's code.

Reviewer: Yes and as I've written, I think the good parts can just be merged, I don't think it's an either/or thing

## Unresolved questions

### Do we want to keep the Masonry brand?

The name Masonry was chosen when that crate wasn't officially part of the Linebender project. Even if we want the codebase and the crate to be added to the Linebender umbrella, its name and general branding are up in the air.

Arguments for:

- "Masonry" is catchy name, easy to pronounce in most european languages.
- The name already exists and the Rust community is vaguely aware of it, to the point is comes up (rarely) in discussions.
- The name evokes brickwork and building stuff out of sturdy rectangle things. It can evoke having a solid foundation that things are built on top of.
- I have [the coolest logo idea ever](https://github.com/PoignardAzur/masonry-rs/issues/7).

Arguments against:

- "A masonry layout" is a thing, especially in web development.
  - People unfamiliar with Rust might come upon this crate and misunderstand what it's for.
  - When talking about how Masonry does layout, discussion could get confusing.

## Future possibilities

> Xilem has added features over Druid that Masonry doesn't support, notably virtual lists and async support.

Reviewer: Well they're prototypes, and as you've already said, they are commented out, and quite a bit out of date/not-working (and I think there may be some possible UX improvements, such as the list callback taking the app state as well as parameter). I think when we're approaching virtualized lists etc. that there will be a refactored version implemented either in xilem as well as a corresponding widget in masonry. A working version is currently on the old [`xilem_tokio`](https://github.com/linebender/druid/tree/xilem_tokio) branch in druid (which is basically the same code as currently in xilem, but working...)

async is an interesting point though, I think long-term async will almost be a requirement for some widgets in masonry. Possibly via a feature gate for async widgets (ideally not hardcoding a specific async runtime)?

- The name already exists and the Rust community is vaguely aware of it, to the point is comes up (rarely) in discussions.
- The name evokes brickwork and building stuff out of sturdy rectangle things. It can evoke having a solid foundation that things are built on top of.
- I have [the coolest logo idea ever](https://github.com/PoignardAzur/masonry-rs/issues/7).

Arguments against:

- "A masonry layout" is a thing, especially in web development.
  - People unfamiliar with Rust might come upon this crate and misunderstand what it's for.
  - When talking about how Masonry does layout, discussion could get confusing.

## Future possibilities

### Performance features

Xilem has added features over Druid that Masonry doesn't support, notably virtual lists and async support.

I think they're currently under-used in Xilem, so Masonry doesn't need to implement them right away for parity.

In the long term, these features will be very important for eventual performance work.

### Directory structure

Xilem's current directory structure is a bit awkward. It looks like:

- `crates/`
  - `xilem_core/`
  - `xilem_web/`
  - (soon) `masonry/`
- `src/`

> In other words, all crates are in a `crates/` directory, except for `xilem/` which is in the root. This makes it seem like eg `xilem_web` is a dependency of `xilem`, when they're disjoint libraries.

Reviewer: We could also move everything to `crates/` (so `crates/xilem`) and make it a pure workspace. Though I don't think the current structure is too bad, especially given that `xilem_web` is still kinda experimental (and either diverges fully, in case a separate repo may make sense, or can be hopefully merged in one way or the other at some point)

Relevant: https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/Moving.20.28or.20splitting.29.20the.20xilem.20crate.3F

An earlier draft of this RFC proposed renaming `xilem` to `xilem_native`, creating a placeholder `xilem` crate to inform about the multiple backends, and moving all of them into the `crates/` folder.

However, these changes were deemed out of scope for this RFC.

> ### Ownership story

Reviewer 3: I think it would be worthwhile to mention that parley is in very much the same state - an individual project that we want to bring under the linebender umbrella.

- Port the list of Github Issues in the `masonry-rs` repository to the `xilem` repository.
- Create a temporary `xilem_masonry` crate, aiming for parity with the current Xilem crate.
- Once parity is reached, replace the current `src/` folder with the contents of `xilem_masonry`. Remove `xilem_masonry/`.
- Publish `xilem`.
- Set the `#masonry` zulip stream's description to a link to the masonry-rs repository.

### How `xilem` source code will be affected

The source code for the `xilem` crate is currently located in the `src/` folder in [the `xilem` repository](https://github.com/linebender/xilem/).

That folder includes:

- **`view/` folder:** Exports reactive-layer Views.
- **`widget/` folder:** Exports platform-layer Widgets.
- **`app.rs`:** Exports `App` struct.
- **`app_main.rs`:** Exports `AppLauncher` struct.
- **`bloom.rs`:** Defines internal `Bloom` struct.
- **`geometry.rs`:** Exports `Axis` struct.
- **`id.rs`:** Exports `Id` type.
- `lib.rs`
- **`text.rs`:** Exports `render_text()` function.

All of this source code except for the `view/` folder and `app.rs` will be removed.

> Xilem will still export a view trait, defined in `src/view/View.rs` using macros from `xilem_core`. Views will be rewritten to have `masonry::Widget` as their `Element` associated type.

Reviewer: ```suggestion
Xilem will still export a view trait, defined in `src/view/view.rs` using macros from `xilem_core`. Views will be rewritten to have `masonry::Widget` as their `Element` associated type.
```

> The `app.rs` composition root will be rewritten to be a minimal wrapper around Masonry's composition root. The Masonry composition root essentially takes callbacks when being built; these callbacks will be piped to the Xilem reactive architecture (building a view tree, reconciliation, and events).

Reviewer: I'm not sure I would go into that detail (callbacks etc.). I think this will clarify in PRs etc. (Because I don't think it will end up just as that, but we'll see).

I think it's enough to just clarify that `app.rs` or `App` (or something along the lines) takes the tasks of building the view tree, reconciliation, and (towards-views) events, and the rest is basically done by masonry which is owned by `app.rs` (however these things communicate, and in whichever threads they may live).
